ui/src/features/project/pipelines/nodes/stage-node.tsx

```diff
@@ -76,15 +80,57 @@ export const StageNode = (props: { stage: Stage }) => {
     onSuccess: (response) => {
       navigate(
         generatePath(paths.promotion, {
-          name: props.stage?.metadata?.namespace,
+          name: projectName,
           promotionId: response.promotion?.metadata?.name
         })
       );
     }
   });
 
-  const totalSubscribersToThisStage =
-    dictionaryContext?.subscribersByStage?.[props.stage?.metadata?.name || '']?.size || 0;
+  const queryFreightMutation = useMutation(queryFreight);
+
+  const checkFreightPromotionEligibility = async (freight: string) => {
+    const freightResponse = await queryFreightMutation.mutateAsync({
+      project: projectName,
+      stage: stageName
+    });
+
+    return Boolean(
+      freightResponse?.groups?.['']?.freight?.find((item) => item?.metadata?.name === freight)
+    );
+  };
+
+  const showManualApproveModal = useManualApprovalModal();
+
+  const ensureEligibilityBeforeAction = async (
+    freight: string | undefined,
+    onEligible: (eligibleFreight: string) => void,
+    options?: { runAfterApproval?: (eligibleFreight: string) => void }
+  ) => {
+    if (!freight) {
+      return;
+    }
+
+    const isEligible = await checkFreightPromotionEligibility(freight);
+
+    if (!isEligible) {
+      showManualApproveModal({
+        freight,
+        projectName,
+        stage: stageName,
+        onApprove: options?.runAfterApproval
+          ? () => {
+              options.runAfterApproval?.(freight);
+            }
+          : undefined
```

Comment:
This condition is not really helpful. Nothing wrong if you redirect the user here, and simplify the code.